content/githubblog/2023-02-28-how-to-build-a-consistent-workflow-for-development-and-operations-teams.md

```diff
@@ -0,0 +1,127 @@
+<p>しかし、これをさらに進化させることができるのですTerraformでは、変更を本番環境に反映させる前に、<code>terraform</code> <code>planという</code>コマンドを使って変更の影響を把握するのが一般的です。</p>
+<h3 id="reviewing-environment-changes-in-a-pull-request">プルリクエストで環境の変更を確認する<a href="#reviewing-environment-changes-in-a-pull-request" class="heading-link pl-2 text-italic text-bold" aria-label="Reviewing environment changes in a pull request"></a></h3>
+<p>HashiCorp は、<a href="https://developer.hashicorp.com/terraform/tutorials/automation/github-actions">GitHub Actions を使って Terraform を自動化</a>する例を示しています。この例では、GitHub Actionsを使ってTerraform Cloudによるリリースをオーケストレーションしています。この例では、<code>terraform plan</code>コマンドの出力をプルリクエストにコピーして承認します（これも、選択した開発フローに依存します）。</p>
+<h3 id="reviewing-environment-changes-using-github-actions-environments">GitHub Actions の環境を使った環境変更のレビュー<a href="#reviewing-environment-changes-using-github-actions-environments" class="heading-link pl-2 text-italic text-bold" aria-label="Reviewing environment changes using GitHub Actions environments"></a></h3>
+<p><a href="https://www.hashicorp.com/blog/automate-infrastructure-provisioning-workflows-with-the-github-action-for-terraform">HashiCorpの</a>例をもとに、もうひとつの例を考えてみましょう。GitHub Actions には、environment (<a href="https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment">環境</a>) という概念が組み込まれています。この環境は、目標とするデプロイ先への論理的なマッピングだと考えてください。保護ルールを環境と関連づけることで、デプロイ前に承認が得られるようになります。</p>
+<p>では、GitHub Action のワークフローを作りましょう。2 つの環境 (計画用とデプロイ用の 2 つ) を用意します。</p>
+<pre><code>name: 'Review and Deploy to EnvironmentA' (レビューと環境 A へのデプロイ)
+を実行します。[push] (プッシュ)
+
+ジョブ
+  レビュー
+    name: 'Terraform Plan' (テラフォームプラン)
+    環境: 環境_a_plan
+    ランズオン: ubuntu-latest
+
+    ステップ
+      - name: 'チェックアウト'
+        用途: アクション/チェックアウト@v2
+
+      - name: 'Terraform Setup' (テラフォームのセットアップ)
+        uses: hashicorp/setup-terraform@v2
+        を使っています。
+          cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }} とします。
+
+      - name: 'Terraform Init'
+        run: テラフォームの初期化
+
+
+      - name: 'Terraform Format' (テラフォームフォーマット)
+        run: テラフォーム fmt -check
+
+      - name: 'Terraform Plan' (テラフォームプラン)
+        run: テラフォーム計画 -input=false
+
+  をデプロイします。
+    名前: 'Terraform'
+    環境: environment_a_deploy
+    実行環境: ubuntu-latest
+    が必要です。[レビュー].
+
+    ステップ
+      - name: 'チェックアウト'
+        用途: アクション/チェックアウト@v2
+
+      - name: 'Terraform Setup' (テラフォームのセットアップ)
+        uses: hashicorp/setup-terraform@v2
+        を使っています。
+          cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }} とします。
+
+      - name: 'Terraform Init'
+        run: テラフォームの初期化
+
+      - name: 'Terraform Plan' (テラフォームプラン)
+        run: Terraform apply -auto-approve -input=false
+</code></pre>
```

Comment:
```suggestion
\ ```
name: 'Review and Deploy to EnvironmentA'
on: [push]

jobs:
  review:
    name: 'Terraform Plan'
    environment: environment_a_plan
    runs-on: ubuntu-latest

    steps:
      - name: 'Checkout'
        uses: actions/checkout@v2

      - name: 'Terraform Setup'
        uses: hashicorp/setup-terraform@v2
        with:
          cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }}

      - name: 'Terraform Init'
        run: terraform init


      - name: 'Terraform Format'
        run: terraform fmt -check

      - name: 'Terraform Plan'
        run: terraform plan -input=false

  deploy:
    name: 'Terraform'
    environment: environment_a_deploy
    runs-on: ubuntu-latest
    needs: [review]

    steps:
      - name: 'Checkout'
        uses: actions/checkout@v2

      - name: 'Terraform Setup'
        uses: hashicorp/setup-terraform@v2
        with:
          cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }}

      - name: 'Terraform Init'
        run: terraform init

      - name: 'Terraform Plan'
        run: terraform apply -auto-approve -input=false
\```
```